util/patch/patch.go

```diff
@@ -86,49 +83,166 @@ func (h *Helper) Patch(ctx context.Context, resource runtime.Object) error {
+// Conditions are ignored at this stage and handled separately.
+func (h *Helper) patchStatus(ctx context.Context, resource runtime.Object) error {
+	before, err := extractStatusWithoutConditions(h.before)
 	if err != nil {
 		return err
 	}
-	if ok {
-		hasStatus = true
-		// if the resource contains a status remove it from our unstructured copy
-		// to avoid uneccsary patching.
-		unstructured.RemoveNestedField(after, "status")
+	after, err := extractStatusWithoutConditions(h.after)
+	if err != nil {
+		return err
 	}
+	if !shouldPatch(before, after) {
+		return nil
+	}
+	return h.client.Status().Patch(ctx, resource.DeepCopyObject(), h.statusPatch)
+}
 
-	var errs []error
+// patchStatusConditions issues a patch if there are any changes to the conditions slice under
+// the status subresource. This is a special case and it's handled separately given that
+// we allow different controllers to act on conditions of the same object.
+//
+// This method has an internal backoff loop. When a conflict is detected, the method
+// asks the Client for the a new version of the resource we're trying to patch.
+//
+// Condition changes are then applied to the latest version of the object, and if there are
+// no unresolvable conflicts, the patch is sent again.
+func (h *Helper) patchStatusConditions(ctx context.Context, resource runtime.Object) error {
+	// Store the patch from the before/after object.
+	conditionPatch := conditions.NewPatch(
+		conditions.UnstructuredGetter(h.before),
+		conditions.UnstructuredGetter(h.after),
+	)
+
+	// Return early if there are no changes.
+	if conditionPatch.IsZero() {
+		return nil
+	}
 
-	if !reflect.DeepEqual(h.before, after) {
-		// only issue a Patch if the before and after resources (minus status) differ
-		if err := h.client.Patch(ctx, resource.DeepCopyObject(), h.resourcePatch); err != nil {
-			errs = append(errs, err)
-		}
+	// Make a copy of the object and store the key used if we have conflicts.
+	resourceCopy := resource.DeepCopyObject()
+	key, err := client.ObjectKeyFromObject(resourceCopy)
+	if err != nil {
+		return err
+	}
+
+	// Define and start a backoff loop to handle conflicts
+	// between controllers working on the same object.
+	//
+	// This has been copied from https://github.com/kubernetes/kubernetes/blob/release-1.16/pkg/controller/controller_utils.go#L86-L88.
+	backoff := wait.Backoff{
+		Steps:    5,
+		Duration: 100 * time.Millisecond,
+		Jitter:   1.0,
 	}
 
-	if (h.hasStatus || hasStatus) && !reflect.DeepEqual(h.beforeStatus, afterStatus) {
-		// only issue a Status Patch if the resource has a status and the beforeStatus
-		// and afterStatus copies differ
-		if err := h.client.Status().Patch(ctx, resource.DeepCopyObject(), h.statusPatch); err != nil {
-			errs = append(errs, err)
+	// Start the backoff loop and return errors if any.
+	return wait.ExponentialBackoff(backoff, func() (bool, error) {
+		// Issue a patch.
+		err := h.client.Status().Patch(ctx, resourceCopy, h.conditionsPatch)
```

Comment:
If I'm not wrong, this first patch is created with an optimistic lock, so it will work only if we are not patching the object or the status before this call.

That means that if we want to support patching conditions for objects that are not setter/cluster API objects we should make the use of optimistic lock optional using an option

Reply:
ptal this has been reworked and forgot to push it

Reply:
IMO the comment is still valid

Reply:
with the latest changes, this is not a problem anymore (patch conditions apply only for setter/cluster API objects, while for other objects conditions remain part of status)